clang/lib/AST/ASTContext.cpp

```diff
@@ -1681,6 +1681,12 @@ const llvm::fltSemantics &ASTContext::getFloatTypeSemantics(QualType T) const {
     if (getLangOpts().OpenMP && getLangOpts().OpenMPIsTargetDevice)
       return AuxTarget->getFloat128Format();
     return Target->getFloat128Format();
+  case BuiltinType::DecimalFloat32:
+    return Target->getDecimalFloat32Format();
+  case BuiltinType::DecimalFloat64:
+    return Target->getDecimalFloat64Format();
+  case BuiltinType::DecimalFloat128:
+    return Target->getDecimalFloat128Format();
```

Comment:
Do these really need to be changeable by the target? Un like the basic float/double/long double types, the _DecimalN types can only legally correspond to the IEEE 754-types, and the only real question is if we enable support for these types on a target (which I don't know the mechanism for in clang).

Reply:
I agree with Joshua in principle here, but a non-target-dependent design would deviate from the overall pattern for every other FP type. I don't think it is worth it. One rather hackish thing we could do would be to declare the `DecimalFloatXXFormat` variables and corresponding `getDecimalFloatXXFormat()` member functions as `static`. That would prevent any possible target deviation while still keeping all the FP support colocated. I don't think doing so is much of a win though, so I'm inclined to just go with what Shafik has here.